fedmsg/crypto/x509_ng.py

```diff
@@ -19,256 +19,221 @@
+    # since the ultimate goal is for it to be obsoleted:
+    # https://mail.python.org/pipermail/cryptography-dev/2017-June/000774.html
+    from cryptography import x509
+    from cryptography.exceptions import InvalidSignature
+    from cryptography.hazmat.backends import default_backend
+    from cryptography.hazmat.primitives import asymmetric, hashes, serialization
+    from OpenSSL.crypto import (X509Store, X509StoreContext, X509StoreContextError,
+                                load_certificate, load_crl, FILETYPE_PEM, X509StoreFlags)
+    _cryptography = True
+except ImportError:  # pragma: no cover
+    _cryptography = False
+
+from . import utils
 import fedmsg.crypto
 import fedmsg.encoding
 
-import logging
-log = logging.getLogger(__name__)
-
-import cryptography.x509
-from cryptography.hazmat.backends import default_backend
-from cryptography.hazmat.primitives import asymmetric, hashes, serialization
-
-import six
 
-if six.PY3:
-    long = int
+_log = logging.getLogger(__name__)
 
 
 def sign(message, ssldir=None, certname=None, **config):
-    """ Insert two new fields into the message dict and return it.
+    """Insert two new fields into the message dict and return it.
 
     Those fields are:
 
         - 'signature' - the computed RSA message digest of the JSON repr.
         - 'certificate' - the base64 X509 certificate of the sending host.
-    """
 
+    Arg:
+        message (dict): An unsigned message to sign.
+        ssldir (str): The absolute path to the directory containing the SSL certificates to
+            use.
+        certname (str): The name of the key pair to sign the message with. This corresponds to
+            the filenames within ``ssldir`` sans prefixes. The key pair must be named
+            ``<certname>.key`` and ``<certname>.crt``
+    Returns:
+        dict: The signed message.
+    """
     if ssldir is None or certname is None:
         error = "You must set the ssldir and certname keyword arguments."
         raise ValueError(error)
 
     message['crypto'] = 'x509'
 
-    with open("%s/%s.crt" % (ssldir, certname), "rb") as f:
-        data = f.read()
-        cryptography.x509.load_pem_x509_certificate(
-            data,
-            default_backend()
-        )
-        # TODO: it seems that there's no way in cryptography to
-        #  get back the PEM block of loaded certificate
-        cert_pem = data[data.decode('utf-8').find('-----BEGIN'):]
-
-    # Opening this file requires elevated privileges in stg/prod.
     with open("%s/%s.key" % (ssldir, certname), "rb") as f:
         rsa_private = serialization.load_pem_private_key(
             data=f.read(),
             password=None,
             backend=default_backend()
         )
 
-    signer = rsa_private.signer(
+    signature = rsa_private.sign(
+        fedmsg.encoding.dumps(message).encode('utf-8'),
         asymmetric.padding.PKCS1v15(),
-        hashes.SHA1()
+        hashes.SHA1(),
```

Comment:
While it's fine right now, it would be great if we could start migrating to new hashes, like sha256/sha512.
Maybe add an indicator which is used before the next version, so we can switch easier in the future?